### ISSUE TITLE
VQC works with COBYLA but not ADAM

### ISSUE THREAD

**Issue by DPeddireddy** [Qiskit/qiskit-machine-learning#446]
### Environment

- **Qiskit Machine Learning version**: 0.4.0
- **Python version**: 3.7.13
- **Operating system**:  Executed on Google Colab

### What is happening?

I've tried to implement a classification problem on Iris dataset using the VQC function which works perfectly well when I set the optimizer to COBYLA but when I change it to ADAM keeping everything else the same, the program doesn't doesn't end when I run fit.

### How can we reproduce the issue?

```python
from sklearn.datasets import load_iris
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler,MinMaxScaler
from keras.utils.np_utils import to_categorical 

import qiskit
from qiskit import QuantumCircuit
from qiskit.circuit import ParameterVector
from qiskit.circuit.library import  ZZFeatureMap, TwoLocal
from qiskit.algorithms.optimizers import COBYLA, L_BFGS_B, ADAM
from qiskit.utils import QuantumInstance, algorithm_globals

iris = load_iris()
X = iris['data']
y = iris['target']
names = iris['target_names']
feature_names = iris['feature_names']

scaler = MinMaxScaler((-3.142, 3.142))
X_scaled = scaler.fit_transform(X)

X_train, X_test, y_train, y_test = train_test_split(
X_scaled, y, test_size=0.2, random_state=2)
y_train = to_categorical(y_train, num_classes=3)
y_test = to_categorical(y_test, num_classes = 3)

backend = Aer.get_backend('qasm_simulator')
backend_options = {"method": "statevector"}
seed = 88
quantum_instance = QuantumInstance(backend, shots=1024,seed_simulator = seed, seed_transpiler = seed,
                                   backend_options=backend_options)

vqc = VQC(
    feature_map=ZZFeatureMap(4),
    ansatz=TwoLocal(4, ['rx', 'ry','rz'], 'cx', entanglement = 'circular', reps = 1, insert_barriers = True),
    loss="cross_entropy",
    #optimizer= COBYLA(maxiter=30),
    optimizer = ADAM(maxiter=30, lr=1e-3, tol = 1e-6),
    quantum_instance=quantum_instance
)

vqc.fit(X_train, y_train)
```

### What should happen?

Program works fine when the optimizer is set to COBYLA but doesn't doesn't end when it's set to ADAM

**Comment by adekusar-drl** [comment]
The script works, but unfortunately slow. The problem is that ADAM is a gradient based optimizer, while COBYLA is gradient free. That means, ADAM estimates gradient with respect to ansatz parameters. For each sample in the dataset, for each parameter there are two additional circuits are executed. In your case it is 120 * 24 * 2 = 5760 circuits. And the gradient is estimated on each iteration. COBYLA does not do that, it just evaluates an objective function defined by `VQC` one time per iteration, thus one circuit per iteration. See the numbers.

**Comment by DPeddireddy** [comment]
> The script works, but unfortunately slow. The problem is that ADAM is a gradient based optimizer, while COBYLA is gradient free. That means, ADAM estimates gradient with respect to ansatz parameters. For each sample in the dataset, for each parameter there are two additional circuits are executed. In your case it is 120 * 24 * 2 = 5760 circuits. And the gradient is estimated on each iteration. COBYLA does not do that, it just evaluates an objective function defined by `VQC` one time per iteration, thus one circuit per iteration. See the numbers.

That makes sense, thank you. If the gradients are being computed using the parameter shift rule, the runtime could probably be improved by computing the gradients parallelly.

**Comment by adekusar-drl** [comment]
@DPeddireddy The problem is more generic, the runtimes may help, but not that much. Here is the cause of this issue: https://github.com/Qiskit/qiskit-terra/issues/7676.
